Make GenLazyNativeFuncDefinition abstract and extensible

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76180
* __->__ #75343

- break up monolithic block of code into more readable functions
- Rename LazyIr generator classes for consistency
- clean up dead code

Differential Revision: [D35455171](https://our.internmc.facebook.com/intern/diff/D35455171)

wconstab committed:
Make GenLazyNativeFuncDefinition abstract and extensible

- Rename LazyIr generator classes for consistency
- clean up dead code

[ghstack-poisoned]

wconstab commented on tools/codegen/dest/lazy_ir.py:
```diff
@@ -168,44 +168,47 @@ class {schema.node_name} : public {self.node_base} {{
+            elif isinstance(arg.lazy_type, BaseCType):
+                if arg.lazy_type.type is tensorListValueT:
+                    lazy_tensor_decls.append(
+                        f"auto lazy_{arg.name}_tensorlist = torch::lazy::GetTensorList({arg.name});")
```
@bdhirsh so in particular, there are a bunch of functions calls here that XLA wants to modify (GetTensorList, GetLtcTensorOrCreateForWrappedNumber, TryGetLtcTensor, ...)

It is possible this is just a short term need, as XLA could migrate to fully using these core APIs but wants to start off using their versions first.

We could also solve this particular problem by making something like a 'TensorFactory' interface and having Xla override that, then making codegen code just call the factory API.  This is my preferred solution in this case, but i thought it would be easier in the short term to unblock XLA using the flexible-codegen approach.

bdhirsh replied:
Ah yeah - so at least in this case, it sounds like the extra extensibility is really just a temporary thing for migration, and not a long term difference between the TS/XLA backends.

TBH, if XLA doesn't want to use the existing `torch::lazy::*` API's yet, I think your `TensorFactory` interface idea sounds like it might be less work compared to building the overridable codegen, and coordinating unwinding it later on (that probably depends on how many / what sort of functions are being extended though). If you think the divergence just a temporary thing though, I think either option sounds ok.

wconstab replied:
yea, i might be convinced in this case.

I think i still want to land this PR which actually does not make anything extensible but just refactors a bit.  I think i like the idea of making GenLazyNativeFunctionDef have a number of smaller methods and then having the __call__ just drop the result of each method into its place in the template.

wconstab replied:
@bdhirsh wdyt about landing this as is (with a warning I added)?

I talked to @JackCaoG and it seems like there is a chicken-egg problem bc they haven't adopted LazyTensor class yet, which makes it harder for them to use our longer term API solution, but easier to get started by overriding the codegen.

bdhirsh replied:
@wconstab `GenLazyNativeFuncDefinition` has a bunch stuff on it, and making the whole class abstract makes it kinda hard to reason about what functions a backend might be overriding or not (technically, someone else could override `GenLazyNativeFuncDefinition.__call__` which would be pretty weird).

What do you think of taking the set of things that XLA should be able to override, and moving them into a different extendible class somewhere (that gets called by this class)? That way XLA won't need to (or have the freedom to) override the entire `lazy_tensor_decls()` function, which is probably too much extensibility anyway. They'd have an individual method for e.g. `BackendExtensions.getTensorListAPI()`

wconstab commented:
@wconstab has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35455171).

wconstab committed:
Update on "Make GenLazyNativeFuncDefinition abstract and extensible"


- Rename LazyIr generator classes for consistency
- clean up dead code

Differential Revision: [D35455171](https://our.internmc.facebook.com/intern/diff/D35455171)

[ghstack-poisoned]

wconstab commented:
@wconstab has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35455171).

bdhirsh commented on tools/setup_helpers/generate_code.py:
```diff
@@ -197,7 +197,7 @@ def main() -> None:
         from tools.codegen.gen_lazy_tensor import run_gen_lazy_tensor
-        from tools.codegen.dest.lazy_ir import TSLazyIR
+        from tools.codegen.dest.lazy_ir import GenTSLazyIR
         run_gen_lazy_tensor(aten_path=aten_path,
```
Just looking around - is `gen_ts_lowerings` still used anywhere? it looks like in `gen_lazy_tensor.py`, it's  used to implicitly reference `GenTSLazyIR`, but now we explicitly pass that class in here instead. 

Do you also need a way to tell the codegen to override the `GenLazyNativeFuncDefinition` abstract class? (Maybe we should also try to bundle them together so there's one place where the extensions live).

bdhirsh reviewed (approved):
Left some minor comments, but approving pre-emptively to unblock! (Feel free to handle them in a followup).

It sounds like the extra flexibility should help XLA migrate their codegen faster in the short term. And we'll be able to remove the abstract base class later once XLA switches over to  later XLA will switch to `torch::lazy::` API's. Hopefully the abstract `GenLazyNativeFuncDefinition` class won't stick around for too long :) but LGTM!

wconstab committed:
Update on "Make GenLazyNativeFuncDefinition abstract and extensible"


- Rename LazyIr generator classes for consistency
- clean up dead code

Differential Revision: [D35455171](https://our.internmc.facebook.com/intern/diff/D35455171)

[ghstack-poisoned]

wconstab committed:
Update on "Make GenLazyNativeFuncDefinition abstract and extensible"


- Rename LazyIr generator classes for consistency
- clean up dead code

Differential Revision: [D35455171](https://our.internmc.facebook.com/intern/diff/D35455171)

[ghstack-poisoned]

wconstab commented:
@wconstab has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35455171).
